File: 알고리즘 기본 구현/CPP/Eratosthenes.cpp
#include <iostream>
#include <vector>
#include <algorithm>

// Faster
std::vector<int> getPrimes(int last) {
    if (last < 2) {
        return std::vector<int>();
    }

    std::vector<bool> isPrime(last + 1, true);
    isPrime[0] = isPrime[1] = false;
    std::vector<int> primes;

    for (int div = 2; div*div <= last; ++div) {
        if (isPrime[div]) {
            for (int i = div*div; i <= last; i += div) {
                isPrime[i] = false;
            }
        }
    }

    for (int i = 2; i < last + 1; ++i) {
        if (isPrime[i]) {
            primes.push_back(i);
        }
    }

    return primes;
}

// Slower
std::vector<int> getPrimes(int first, int last) {
    first = std::max(first, 2);
    int count = last - first + 1;
    std::vector<bool> isPrime(count, true);
    std::vector<int> primes;

    for (int div = 2; div*div <= last; ++div) {
        int fillStart = -1;
        for (int i = std::max(0, div - first); i < count; ++i) {
            if (isPrime[i] && (first + i) % div == 0) {
                fillStart = first + i;
                break;
            }
        }

        if (fillStart != -1) {
            for (int i = std::max(div*div, fillStart) - first; i < count; i += div) {
                isPrime[i] = false;
            }
        }
    }

    for (int i = 0; i < count; ++i) {
        if (isPrime[i]) {
            primes.push_back(i + first);
        }
    }

    return primes;
}

File: 알고리즘 기본 구현/Rust/Eratosthenes.rs
fn primes(last: usize) -> Vec<usize> {
    if last < 2 {
        return vec![];
    }

    let mut is_prime = vec![true; last + 1];
    is_prime[0] = false;
    is_prime[1] = false;
    let mut div = 2;

    while div * div <= last {
        if is_prime[div] {
            for i in (div * div..=last).step_by(div) {
                is_prime[i] = false;
            }
        }
        div += 1;
    }

    is_prime
        .into_iter()
        .enumerate()
        .skip(2)
        .filter_map(|(x, p)| if p { Some(x) } else { None })
        .collect()
}